eclass/dub.eclass

```diff
@@ -0,0 +1,305 @@
+
+	edo dub --verbose "${@}"
+}
+
+# @ECLASS_VARIABLE: DUB_LOCAL_REGISTRY
+# @DESCRIPTION:
+# A directory in which dub package archives can be placed for dub to be
+# able to fetch them in the future.  This is useful only if dub _must_
+# perform a fetch operation, perhaps as part of tests, in all other
+# cases DUB_DEPENDENCIES should be used instead.
+#
+# You can use dub_copy_dependencies_locally as a convenience function.
+#
+# Example:
+# @CODE
+# src_unpack() {
+#   dub_src_unpack
+#   use test && dub_copy_dependencies_locally "my_packages@1.1.0"
+# }
+# src_test() {
+#   dub_src_test
+#
+#   # additional test that performs `dub fetch my_package@1.1.0`
+# }
+# @CODE
+DUB_LOCAL_REGISTRY="${EDUB_HOME}/local_registry"
+
+# @FUNCTION: dub_copy_dependencies_locally
+# @USAGE: [dependencies]...
+# @DESCRIPTION:
+# Copy the given dependencies to ${DUB_LOCAL_REGISTRY}.  This function
+# can only be called inside src_unpack.
+#
+# The format of the arguments is the same as the format of
+# ${DUB_DEPENDENCIES}.
+dub_copy_dependencies_locally() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	[[ ${EBUILD_PHASE} == unpack ]] \
+		|| die "${FUNCNAME} must only be called inside src_unpack"
+
+	local dep
+	for dep; do
+		einfo "Making ${dep} available locally"
+		cp "${DISTDIR}/${dep/@/-}.zip" "${DUB_LOCAL_REGISTRY}" || die
+	done
+}
+
+
+# @FUNCTION: dub_gen_settings
+# @DESCRIPTION:
+# Generate a settings.json file for dub.
+dub_gen_settings() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	mkdir -p "${EDUB_HOME}" || die "Could not create dub home directory"
+	mkdir -p "${DUB_LOCAL_REGISTRY}" || die "Could not create dub registry directory"
+
+	# dub merges the settings from all configuration files so settings
+	# which greatly change the build need to be set to a safe
+	# value. Unfortunately the 'customCachePaths' and 'registryUrls'
+	# settings are cumulative, i.e. they will append to the previous
+	# value, not overwrite it.
+	cat > "${EDUB_HOME}/settings.json" <<-EOF || die "Failed to create dub settings.json file"
+	{
+		"registryUrls": [
+			"file://${DUB_LOCAL_REGISTRY}"
+		],
+		"skipRegistry": "standard",
+
+		"customCachePaths": [],
+		"defaultCompiler": "",
+		"defaultArchitecture": ""
+	}
+	EOF
+
+	export DUB_HOME="${EDUB_HOME}"
+}
+
+# @FUNCTION: dub_src_unpack
+# @DESCRIPTION:
+# Unpack all sources and generate a dub configuration file
+dub_src_unpack() {
+	dub_gen_settings
+
+	local name ver dep
+	local -A dubFiles
+	for dep in "${DUB_DEPENDENCIES[@]}"; do
+		name="${dep%@*}"
+		ver="${dep#*@}"
+		dubFiles["${name}-${ver}.zip"]="${name}"
+	done
+
+	local dubFetch=(
+		dub
+		--skip-registry=all
+		--registry="file://${DISTDIR}"
+		fetch
+	)
+	local file
+	for file in ${A}; do
+		if [[ ${dubFiles[${file}]} ]]; then
+			name="${dubFiles[${file}]}"
+			ver="${file#${name}-}"
+			ver="${ver%.zip}"
+
+			edo "${dubFetch[@]}" "${name}@${ver}"
+
+			# If it's the main package
+			[[ ${file} == ${PN}-$(ver_rs 3 - 4 .)* ]] \
```

Comment:
This looks a bit gnarly and maybe a comment explaining why it's right wouldn't go amiss, but it might be obvious to someone who works with D a lot, so I don't insist.